### PR TITLE
Makes damaging items stop on hit and play a sound

### PR DESCRIPTION
**Pull request by Bright0**
## About the PR
Fixes: #3920
Previously, thrown items just kinda passed through everything silently, even ones that damage things. This PR mimics behavior in SS13, where thrown items will stop in their tracks on collision and play a hit sound depending on what the item is. This also makes it so that thrown items should only collide and deal damage one time and then stop, but there's room for extra checks to be made if an item shouldn't care. This will also help illuminate bugs with the throwing system now that there's an audio cue when something's hit.

**Screenshots**
https://user-images.githubusercontent.com/55061890/201219384-85e0f960-c9dc-4e7e-ae6a-4aeca21f1223.mp4

**Changelog**
:cl:
- fix: thrown items that deal damage don't phase through people
- add: thrown items that deal damage make a sound on hit

### CODE REVIEW

**Comment by metalgearsloth** [comment]
Even if I wanted this this is just the restitution field.

**Comment by Bright0** [comment]
> Even if I wanted this this is just the restitution field.

I don't see the issue. I get you're saying this is a bad thing but not why you feel that way

**Comment by metalgearsloth** [comment]
> > Even if I wanted this this is just the restitution field.
> 
> I don't see the issue. I get you're saying this is a bad thing but not why you feel that way

Being able to bounce items around corners is far more interesting gameplay-wise than setting velocity to 0, especially in weightless settings.

See also https://github.com/space-wizards/space-station-14/blob/4ef144d865ea6e4193efd2789940cd9677066551/Resources/Prototypes/Entities/Objects/base_item.yml#L33

**Comment by Bright0** [comment]
> Being able to bounce items around corners is far more interesting gameplay-wise than setting velocity to 0
I wasn't aware the collision triggered for every collision, lemme fix the logic a bit. at the very least though, it needs to set velocity to 0 and prevent further collisions from a throw in certain situations like throwing glass shards at people

**Comment by Bright0** [comment]
so in it's current state, the code performs as intended and is much less hacky. Items with a DamageOtherOnHit component can be specified to stop when it hits something via YAML, and they will by default. DamageOtherOnHitSystem asks ThrownItemSystem to stop the item from moving, but I don't know a clean way to make it stop raising collision events after the first hit because currently objects may still hit twice even after asking it to land.

File: Content.Shared/Throwing/LandEvent.cs [hunk 1]
```diff
@@ -14,8 +14,14 @@ public sealed class LandEvent : EntityEventArgs
     /// <summary>
     /// Raised when a thrown entity is no longer moving.
     /// </summary>
-    public sealed class StopThrowEvent : EntityEventArgs
+    [ByRefEvent]
+    public struct StopThrowEvent
```
Comment (AJCM-git) [inline Content.Shared/Throwing/LandEvent.cs:18]: make it a record struct

File: Content.Shared/Throwing/ThrowEvents.cs [hunk 2]
```diff
@@ -31,20 +33,52 @@ public ThrowEvent(EntityUid? user, EntityUid thrown, EntityUid target)
     /// <summary>
     ///     Raised directed on the target entity being hit by the thrown entity.
     /// </summary>
-    public sealed class ThrowHitByEvent : ThrowEvent
+    [ByRefEvent]
+    public struct ThrowHitByEvent
```
Comment (AJCM-git) [inline Content.Shared/Throwing/ThrowEvents.cs:37]: same

File: Content.Shared/Throwing/ThrowEvents.cs [hunk 3]
```diff
@@ -31,20 +33,52 @@ public ThrowEvent(EntityUid? user, EntityUid thrown, EntityUid target)
...
         }
     }
 
     /// <summary>
-    ///     Raised directed on the thrown entity that hits another.
+    ///     Raised directed on the thrown entity that hits another. 'User' is whoever threw it, if anyone.
     /// </summary>
-    public sealed class ThrowDoHitEvent : ThrowEvent
+    [ByRefEvent]
+    public struct ThrowDoHitEvent
```
Comment (AJCM-git) [inline Content.Shared/Throwing/ThrowEvents.cs:62]: same

File: Content.Shared/Throwing/ThrownItemSystem.cs [hunk 4]
```diff
@@ -22,6 +23,7 @@ public sealed class ThrownItemSystem : EntitySystem
         [Dependency] private readonly SharedContainerSystem _containerSystem = default!;
         [Dependency] private readonly ISharedAdminLogManager _adminLogger = default!;
         [Dependency] private readonly FixtureSystem _fixtures = default!;
+        [Dependency] private readonly EntityManager _entityManager = default!;
```
Comment (AJCM-git) [inline Content.Shared/Throwing/ThrownItemSystem.cs:26]: dont resolve entity manager, entity system already has a dependency

File: Content.Shared/Throwing/ThrownItemSystem.cs [hunk 5]
```diff
@@ -114,18 +119,25 @@ private void StopThrow(EntityUid uid, ThrownItemComponent thrownItemComponent)
                 }
             }
 
-            EntityManager.EventBus.RaiseLocalEvent(uid, new StopThrowEvent {User = thrownItemComponent.Thrower}, true);
+            var stopThrow = new StopThrowEvent(thrownItemComponent.Thrower);
+            EntityManager.EventBus.RaiseLocalEvent(uid, ref stopThrow, true);
```
Comment (AJCM-git) [inline Content.Shared/Throwing/ThrownItemSystem.cs:123]: use  the proxy method RaiseLocalEvent() , do the same for the rest of things calling EntityManager in here

File: Content.Shared/Throwing/ThrownItemSystem.cs [hunk 6]
```diff
@@ -114,18 +119,25 @@ private void StopThrow(EntityUid uid, ThrownItemComponent thrownItemComponent)
...
+        /// <summary>
+        ///     Called when thrownItem should hit the ground and slide, can stop movement entirely if specified.
+        /// </summary>
+        public void LandComponent(ThrownItemComponent thrownItem, bool stopMoving = false)
         {
             if (thrownItem.Deleted || Deleted(thrownItem.Owner) || _containerSystem.IsEntityInContainer(thrownItem.Owner)) return;
 
             var landing = thrownItem.Owner;
 
+            if (stopMoving && _entityManager.TryGetComponent(landing, out IPhysBody? physics))
```
Comment (AJCM-git) [inline Content.Shared/Throwing/ThrownItemSystem.cs:136]: proxy method

File: Content.Server/Damage/Systems/DamageOtherOnHitSystem.cs [hunk 7]
```diff
@@ -4,26 +4,41 @@
...
+        private void OnDoHit(EntityUid uid, DamageOtherOnHitComponent component, ref ThrowDoHitEvent args)
         {
             var dmg = _damageableSystem.TryChangeDamage(args.Target, component.Damage, component.IgnoreResistances, origin: args.User);
 
             // Log damage only for mobs. Useful for when people throw spears at each other, but also avoids log-spam when explosions send glass shards flying.
             if (dmg != null && HasComp<MobStateComponent>(args.Target))
                 _adminLogger.Add(LogType.ThrowHit, $"{ToPrettyString(args.Target):target} received {dmg.Total:damage} damage from collision");
+
+            // Only play hitsound of thrown item and signal to stop when hitting mobs
+            if (HasComp<MobStateComponent>(args.Target))
```
Comment (AJCM-git) [inline Content.Server/Damage/Systems/DamageOtherOnHitSystem.cs:31]: doesnt the check above already  check for mob state? why not unify this

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Shared/Throwing/ThrownItemSystem.cs (hunk 6 above)
Comment (Bright0) [reply Content.Shared/Throwing/ThrownItemSystem.cs]: I _think_ I used proxy methods everywhere I could with the new commits, can't seem to just do TryGetComponent on its own, got rid of resolving EntityManager but it still needs EntityManager.TryGetComponent in this particular instance

Comment (AJCM-git) [reply Content.Shared/Throwing/ThrownItemSystem.cs]: You can use TryComp<>() proxy method

**Review by AJCM-git — CHANGES_REQUESTED** [review]
merge master

File: Content.Server/Damage/Systems/DamageOtherOnHitSystem.cs [hunk 8]
```diff
@@ -4,26 +4,41 @@
...
+
+                // Log nonzero damage only for mobs. Useful for when people throw spears at each other, but also avoids log-spam when explosions send glass shards flying.
+                if (dmg != null)
+                    _adminLogger.Add(LogType.ThrowHit, $"{ToPrettyString(args.Target):target} received {dmg.Total:damage} damage from collision");
+
+                if (component.StopOnHit)
+                {
+                    args.StopCollisions = true;
+                    args.StopMoving = true;
+                }
```
Comment (metalgearsloth) [inline Content.Server/Damage/Systems/DamageOtherOnHitSystem.cs:39]: This should be an event.

File: Content.Shared/Throwing/ThrowEvents.cs [hunk 9]
```diff
@@ -31,20 +33,52 @@ public ThrowEvent(EntityUid? user, EntityUid thrown, EntityUid target)
...
-        public ThrowHitByEvent(EntityUid? user, EntityUid thrown, EntityUid target) : base(user, thrown, target)
+        public EntityUid? User;
+        public EntityUid Thrown;
+        public EntityUid Target;
+
+        public ThrowHitByEvent(EntityUid? user, EntityUid thrown, EntityUid target)
         {
+            User = user;
+            Thrown = thrown;
+            Target = target;
```
Comment (metalgearsloth) [inline Content.Shared/Throwing/ThrowEvents.cs:47]: As above + merge conflicts.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Needs updating.

**Comment by metalgearsloth** [comment]
@Bright0 are you coming back to this.

**Comment by Bright0** [comment]
> @Bright0 are you coming back to this.

Yes, apologies for the delay, life stuff. still attempting to work out issues with merging master as I'm new to git

**Comment by AJCM-git** [comment]
reopen once you fix your git

**Pull request closed without merging**